client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -768,11 +833,56 @@ private void setHeaders(HttpRequest req, Collection<Header> requestHeaders) {
                     req.addHeader(defaultHeader);
                 }
             }
+            // Add compatibility request headers if compatibility mode has been enabled
+            if (useCompatibility) {
+                addCompatibilityFor(req, entityHeader, "Content-Type");
+                addCompatibilityFor(req, entityHeader, "Accept");
+            }
             if (compressionEnabled) {
                 req.addHeader("Accept-Encoding", "gzip");
             }
         }
 
+        /**
+         * Go through all the request's existing headers, looking for {@code headerName} headers and if they exist,
+         * changing them to use version compatibility. If no request headers are changed, modify the entity type header if appropriate
+         */
+        private void addCompatibilityFor(HttpRequest req, Header entityHeader, String headerName) {
+            // Modify any existing "Content-Type" headers on the request to use the version compatibility, if available
+            boolean contentTypeModified = false;
+            for (Header header : req.getHeaders(headerName)) {
+                contentTypeModified = contentTypeModified || modifyHeader(req, header, headerName);
+            }
+
+            // If there were no request-specific headers, modify the request entity's header to be compatible
+            if (entityHeader != null && contentTypeModified == false) {
+                contentTypeModified = modifyHeader(req, entityHeader, headerName);
+            }
+
+            // If there were no changed headers at all, add the default compatibility header
+            if (contentTypeModified == false) {
+                req.addHeader(headerName, EntityType.APPLICATION_VND_PREFIX + EntityType.JSON.header() + "; compatible-with=7");
+            }
+        }
+
+        /**
+         * Modify the given header to be version compatible, if necessary.
+         * Returns true if a modification was made, false otherwise.
+         */
+        private boolean modifyHeader(HttpRequest req, Header header, String headerName) {
+            for (EntityType type : EntityType.values()) {
```

Comment:
I'm a bit worried about the cost of this method which is on the hot path for every request.

What about:
- adding the two source & target mime-type strings as `EntityType` properties (i.e. `application/json` and `application/vnd.elasticsearch+json; compatible-with=7` instead of just `json`)
- then use a `startsWith()` check instead of `contains`(we cannot just test for equality as there can be a trailing `;charset=utf-8`)
- and finally just set the header value from the `EntyTypes` target property, with no string concatenation.

Reply:
Yes, I'm definitely not thrilled about the cost, my only consolation is that we short-circuit (and the most common entity types are checked first) and that the number of headers/entitytypes is usually small.

> adding the two source & target mime-type strings as EntityType properties (i.e. application/json and application/vnd.elasticsearch+json; compatible-with=7 instead of just json)

Yeah this makes sense rather than constructing them over and over

> then use a startsWith() check instead of contains(we cannot just test for equality as there can be a trailing ;charset=utf-8)

Yeah, I was trying to be paranoid about headers without those things, but looking at the RFC spec it does appear to require the content-type before the charset, so I'll do this.